src/bin/rustup-init.rs

```diff
@@ -25,41 +24,46 @@ use rustup::cli::proxy_mode;
 use rustup::cli::rustup_mode;
 use rustup::cli::self_update;
 use rustup::cli::setup_mode;
+use rustup::currentprocess::{process, with, OSProcess};
 use rustup::env_var::RUST_RECURSION_COUNT_MAX;
 use rustup::utils::utils;
 
 fn main() {
-    if let Err(ref e) = run_rustup() {
-        common::report_error(e);
-        std::process::exit(1);
-    }
+    let process = OSProcess {};
+    with(Box::new(process), || match run_rustup() {
+        Err(ref e) => {
+            common::report_error(e);
+            std::process::exit(1);
+        }
+        Ok(utils::ExitCode(c)) => std::process::exit(c),
```

Comment:
Because exit's occur deep in call stacks, process() can't have an exit() method unless we implement that as panic() - which is ugly, or a macro equivalent try! - which isn't a method - or we get into compiler plugins - which is opaque to readers.

We could downcast errors to extract ExitCode's from errors that have them, but errors can still be not propogated, and downcasting with error-chain is going to just make the migration from that more complicated I feel.

We could treat all errors as 'bad thing exit 1' and Ok(explicit error code) as a form of success, even if that means exiting non-zero e.g. for the proxy mode. Which is what I chose

Thus exit() is implemented as structured return: Err(badthing) for error class returns, and Ok(ExitCode(n)) for deliberate non-error returns where that is needed, which is here.

Reply:
Fair enough.  I suppose we just don't do in-process testing of this part of `rustup-init` and we'll be fine. :+1:

Reply:
Well we can't - its not in the library (it is in bin/ :) ). We could move some of this into the lib, but actually I think this is sufficiently shallow that bugs are remote.